test/sshkit/scp/upload_test.exs

```diff
@@ -0,0 +1,155 @@
+defmodule SSHKit.SCP.UploadTest do
+  use ExUnit.Case, async: true
+  import Mox
+
+  alias SSHKit.SCP.Upload
+  alias SSHKit.SSHMock
+
+  @local "test/fixtures/local_dir"
+  @remote "/home/test/code"
+  @conn %SSHKit.SSH.Connection{}
+  @chan %SSHKit.SSH.Channel{}
+
+  describe "new/3" do
+    test "returns a new upload struct" do
+      upload = Upload.new(@local, @remote)
+      local_expanded = @local |> Path.expand()
+      assert %Upload{local: ^local_expanded, remote: @remote} = upload
+    end
+
+    test "returns a new upload struct with options" do
+      options = [recursive: true]
+      upload = Upload.new(@local, @remote, options)
+      assert %Upload{options: ^options} = upload
+    end
+
+    test "upload struct has initial state" do
+      %Upload{state: state} = Upload.new(@local, @remote)
+      current_directory = @local |> Path.expand() |> Path.dirname()
+      assert state == {:next, current_directory, [["local_dir"]], []}
+    end
+
+    test "upload struct has a handler function" do
+      %Upload{handler: handler} = Upload.new(@local, @remote)
+      assert is_function(handler)
+    end
+  end
+
+  describe "exec/2" do
+    test "returns error when trying to upload a directory non-recursively" do
+      upload = Upload.new(@local, @remote, recursive: false)
+      assert {:error, _msg} = Upload.exec(upload, @conn)
+    end
+
+    test "uses the provided timeout option" do
+      upload = Upload.new(@local, @remote, recursive: true, timeout: 55, ssh: SSHMock)
+      SSHMock |> expect(:run, fn(_, _, timeout: timeout, acc: {:cont, _}, fun: _) ->
+        assert timeout == 55
+        {:ok, :success}
+      end)
+
+      assert {:ok, :success} = Upload.exec(upload, @conn)
+    end
+
+    test "performs an upload" do
+      upload = Upload.new(@local, @remote, recursive: true, ssh: SSHMock)
+      SSHMock |> expect(:run, fn(connection, command, timeout: timeout, acc: {:cont, state}, fun: handler) ->
+        assert connection == @conn
+        assert command == SSHKit.SCP.Command.build(:upload, upload.remote, upload.options)
+        assert timeout == :infinity
+        assert state == upload.state
+        assert handler == upload.handler
+        {:ok, :success}
+      end)
+
+      assert {:ok, :success} = Upload.exec(upload, @conn)
+    end
+  end
+
+  describe "exec handler" do
+    setup do
+      {:ok, upload: Upload.new(@local, @remote), msg: {:data, @chan, 0, <<0>>}}
```

Comment:
The `msg` is pretty much an `ack`nowledgement. I could imagine renaming it could make it's purpose clearer.